src/set.ac

```diff
@@ -0,0 +1,510 @@
+            not self.contains(x)
+        }
+    }
+
+    define c(self) -> Set<K> {
+        Set<K>.new(negate_fun(self.contains))
+    }
+
+    define is_singleton(self) -> Bool {
+        exists(a: K) {
+            self = Set<K>.singleton(a)
+        }
+    }
+}
+
+theorem empty_set_is_empty<K> {
+    (Set<K>.empty_set).is_empty
+}
+
+theorem singleton_set_is_singleton<K>(a: K) {
+    (Set<K>.singleton(a)).is_singleton
+}
+
+theorem compl_of_compl_is_self<K>(s: Set<K>) {
+    s.c.c = s
+} by {
+    forall(x: K) {
+        if s.contains(x) {
+            s.c.c.contains(x)
+        } else {
+            not s.c.c.contains(x)
+        }
+        s.c.c.contains(x) = s.contains(x)
+    }
+
+    s.c.c.contains = s.contains
+}
+
+theorem empty_set_compl_is_universal<K> {
+    (Set<K>.empty_set).c = Set<K>.universal_set
+} by {
+    let e = Set<K>.empty_set
+    let u = Set<K>.universal_set
+
+    forall(x: K) {
+        e.c.contains(x)
+        u.contains(x)
+    }
+
+    e.c.contains = u.contains
+}
+
+theorem universal_set_compl_is_empty<K> {
+    (Set<K>.universal_set).c = Set<K>.empty_set
+} by {
+    let e = Set<K>.empty_set
+    let u = Set<K>.universal_set
+
+    forall(x: K) {
+        not u.c.contains(x)
+        not e.contains(x)
+    }
+
+    u.c.contains = e.contains
+}
+
+theorem singleton_set_is_not_empty<K>(a: K) {
+    not (Set<K>.singleton(a)).is_empty
+} by {
+    let s = Set<K>.singleton(a)
+
+    s.contains(a)
+}
+
+// Subsets
+attributes Set<K> {
+    /// self ⊆ s
+    define subset(self, s: Set<K>) -> Bool {
+        forall(x: K) {
+            self.contains(x) implies s.contains(x)
+        }
+    }
+
+    /// self ⊇ s
+    define superset(self, s: Set<K>) -> Bool {
+        s.subset(self)
+    }
+}
+
+// Subset theorems
+theorem empty_set_is_always_subset<K>(s: Set<K>) {
+    (Set<K>.empty_set).subset(s)
+}
+
+theorem all_sets_subset_universal<K>(s: Set<K>) {
+    s.subset(Set<K>.universal_set)
+}
+
+theorem subset_refl<K>(s: Set<K>) {
+    s.subset(s)
+}
+
+theorem subset_trans<K>(a: Set<K>, b: Set<K>, c: Set<K>) {
+    a.subset(b) and b.subset(c) implies a.subset(c)
+}
+
+theorem double_inclusion<K>(a: Set<K>, b: Set<K>) {
+    a.subset(b) and b.subset(a) implies a = b
+} by {
+    forall(x: K) {
+        a.contains(x) implies b.contains(x)
+        b.contains(x) implies a.contains(x)
+        a.contains(x) = b.contains(x)
+    }
+    a.contains = b.contains
+}
+
+// "Standard" union and intersection of two sets
+define elem_in_union<K>(a: Set<K>, b: Set<K>, x: K) -> Bool {
+    a.contains(x) or b.contains(x)
+}
+
+define elem_in_intersection<K>(a: Set<K>, b: Set<K>, x: K) -> Bool {
+    a.contains(x) and b.contains(x)
+}
+
+// Difference
+define elem_in_difference<K>(a: Set<K>, b: Set<K>, x: K) -> Bool {
+    a.contains(x) and not b.contains(x)
+}
+
+attributes Set<K> {
+    /// self ∪ s
+    define union(self, s: Set<K>) -> Set<K> {
+        Set.new(elem_in_union(self, s))
+    }
+
+    /// self ∩ s
```

Comment:
unicode is cool, I wonder if we want to make it possible for a larger set of characters to be operators, so that we could actually just use these as infix operators. It would be annoying to write though unless there's somehow support for it.